### DISCUSSION TITLE
Popup : set background colour ?

### DISCUSSION THREAD

**Discussion by subv0x** [RomanKubiak/ctrlr#597] [General]
hello folks, 
trying to figure out how to change the background and highlight bar colours in popup menus. 
https://docs.juce.com/master/classPopupMenu.html

but apparently Ctrlr does not have access to this directly, and requires writing a class to hook it up, according to chatGPT, which gave the following:

```
In JUCE PopupMenu, you can set the text color for items using the setColour method 
and specifying TextColourId for text color. Unfortunately, this is not directly available 
in the Lua API for Ctrlr, but you can use a similar approach with a custom LookAndFeel.

Here's how you can modify your code to set the text color for submenus:

lua

-- Custom LookAndFeel for PopupMenu
local CustomLookAndFeel = {}
CustomLookAndFeel.new = function()
    local self = {}
    self.__index = self

    setmetatable(self, {
        __index = LookAndFeel
    })

    return self
end

-- Override item text color
function CustomLookAndFeel:drawPopupMenuItem(g, text, image, isSeparator, width, height, isHighlighted, isSelected)
    local textColour = Colour(0xFFFF0000) -- Set your desired text color here
    g:setColour(textColour)
    g:drawText(text, 4, 0, width - 8, height, JustificationType.left, true)
end

-- Use CustomLookAndFeel for PopupMenu
local customLookAndFeel = CustomLookAndFeel.new()

-- Called when a mouse is down on this component
popupPCM3 = function(comp, event)
    -- This stops issues during panel bootup -- else popup will hang startup!
    if panel:getRestoreState() == true or panel:getProgramState() == true then
        return
    end

    local bankSelect = panel:getModulatorByName("bankSelect"):getValue()
    pcmTables() -- load pcm lists A & B
    local menu, sub1, sub2 = PopupMenu(), PopupMenu(), PopupMenu()

    -- Set the LookAndFeel for the PopupMenu
    menu:setLookAndFeel(customLookAndFeel)
    
    -- add submenus
    menu:addSubMenu("select", sub1, true, Image(), false, 0)
    menu:addSeparator()
    menu:addSubMenu("write", sub2, true, Image(), false, 0)
    menu:addSeparator()

    -- populate submenus
    for i = 1, 128 do
        sub1:addColouredItem(i, i .. ":" .. pcmListA[i], Colour(0xFF0000FF), true, false, Image())
        sub2:addColouredItem(i, i .. ":" .. pcmListB[i], Colour(0xFFFF8800), true, false, Image())
    end

    -- display
    local ret = menu:show(20) -- text size -- ?? parameters? show(0,0,0,0)
    if ret ~= nil then
        -- do something
    end
end

In this example, CustomLookAndFeel is a Lua class that inherits from 
the JUCE LookAndFeel class. It overrides the drawPopupMenuItem method 
to set the text color for items. An instance of this custom look and feel is 
assigned to the main PopupMenu using setLookAndFeel. Submenus 
(sub1 and sub2) inherit this look and feel automatically. You can modify the 
color in the setColour method within drawPopupMenuItem to set your 
desired text color for items.
```

but we can give items a text colour when we create them. what about background colour?
there is this, 'colourIds', with the enumerator: 'backgroundColourId ':
https://docs.juce.com/master/classPopupMenu.html#a0888091ddab0f599bb9d2fab28153f6b

also 'headerTextColourId', 'highlightedBackgroundColourId', 'highlightedTextColourId'. that's what i'm looking for.
there is a link to this: LookAndFeel::setColour
https://docs.juce.com/master/classLookAndFeel.html#a167a1e914771f32e433a0d45aaba45e3

what would that look like?

edit: cannot find mention of 'drawPopupMenuItem' - did chatGPT make that up? just googled it. oh no, here it is: 
https://docs.juce.com/master/structPopupMenu_1_1LookAndFeelMethods.html

**Comment by damiensellier** [comment]
popupmenu in Juce is not a component so the only way to style it is via lookandfeel.
I don't think the whole thing is implemented in the Ctrlr LuaBind library. This is why you cannot pass those particular values to the popup.
I'll see if I can add an extra LUA bind function to add all the setColour to all the component elements.

drawPopupMenuItem exists, it's everywhere in the ctrlr menuBar: 
https://docs.juce.com/master/structPopupMenu_1_1LookAndFeelMethods.html#a65f8fd08fdf0a6c9d6fe7eafa1afb020

**Reply by damiensellier** [reply]
This is what's available to invoke from LUA : 
https://github.com/RomanKubiak/ctrlr/blob/8aa00d82127acda42ad9ac9b7b479461e9436aa4/Source/Lua/JuceClasses/LLookAndFeel.h#L52

void **drawPopupMenuBackground** (Graphics &g, int width, int height);
//void **drawPopupMenuItem** (Graphics &g, const Rectangle<int> &areaInt, bool isSeparator, bool isActive, bool isHighlighted, bool isTicked, bool hasSubMenu, const String &text, const String &shortcutKeyText, const Drawable* drawableIcon, const Colour *textColourPtr);
void **drawPopupMenuUpDownArrow** (Graphics &g, int width, int height, bool isScrollUpArrow);
void **getIdealPopupMenuItemSize** (const String &text, bool isSeparator, int standardMenuItemHeight, int &idealWidth, int &idealHeight);


As you can see drawPopupMenuItem is commented so it's not bound to LUA, I really don't know why it's been removed. I'll uncomment it in a newer version to see if it's working.


EDIT : Maybe it's been commented because of the JUCE 6.0.5 update : 

```
Version 6.0.5
=============

Change
------
New pure virtual methods accepting `PopupMenu::Options` arguments have been
added to `PopupMenu::LookAndFeelMethods`.

Possible Issues
---------------
Classes derived from `PopupMenu::LookAndFeelMethods`, such as custom
LookAndFeel classes, will not compile unless these pure virtual methods are
implemented.

Workaround
----------
The old LookAndFeel methods still exist, so if the new Options parameter is not
useful in your application, your implementation of
`PopupMenu::LookAndFeelMethods` can simply forward to the old methods. For
example, your implementation of `drawPopupMenuBackgroundWithOptions` can
internally call your existing `drawPopupMenuBackground` implementation.

Rationale
---------
Allowing the LookAndFeelMethods to access the popup menu's options allows for
more flexible styling. For example, a theme may wish to query the menu's target
component or parent for colours to use.
```

**Reply by subv0x** [reply]
thanks! i forgot to check the Ctrlr pages.
there it is:
`void drawPopupMenuBackground (Graphics &g, int width, int height);`

so does that mean i can go eg: popup:drawPopupMenuBackground(width,height) ? what is 'Graphics &g'? what does '&g' signify?
googled and got 'Drawable Class'':
https://docs.juce.com/master/classDrawable.html

no mention of a colour value.

and does this mean i don't have to do the metatable thing? ChatGPT was very active in proposing solutions. i wasn't sure how to put it all together. can post it here to see if there's any pointers. (bit over my head tbh) 

the last thing it suggested was to make a single coloured pixel - but this didn't work:

```
    -- Create a small image with the desired color
    local image = Image(Image.RGB, 1, 1, true)
    local g = image:createGraphics(true)
    g:fillAll(Colour(color))
    g:setColour(Colour(0x00000000))  -- Set alpha to 0 for transparency
    g:drawRect(0, 0, 1, 1, 1)

    -- Create a PopupMenu
    local popup = PopupMenu()

    -- Set the background image
    popup:setBackgroundImage(popup:getLookAndFeel():createImageBackground(ImageCache:getImageForHash(image:hashCode())))
```

i'm presenting it with the contents of that page right now, to see what it comes back with  :-) 
(taking rather a long time)

**Reply by subv0x** [reply]
:-D
"ChatGPT

Thank you for providing the LookAndFeel class from Ctrlr. Now that we have this, we can try creating a custom LookAndFeel for the PopupMenu and setting its colors. Please note that the specific functionalities provided by Ctrlr might have some differences from the standard JUCE LookAndFeel, so we might need to adapt accordingly.

Here's an example of how you might modify your LookAndFeel class to customize the PopupMenu: "

**Reply by subv0x** [reply]
ok and this is the code it posted, which is ?? 

```
// Assuming the code is within the LookAndFeelBase class definition

class LookAndFeelBase : public LookAndFeel_V2
{
    // ... other methods ...

    void drawPopupMenuBackground(Graphics& g, int width, int height) override
    {
        g.fillAll(Colours::green); // Set the background color of the PopupMenu to green
    }

    void drawPopupMenuItem(Graphics& g, const Rectangle<int>& area, bool isSeparator, bool isActive, bool isHighlighted, bool isTicked, bool hasSubMenu, const String& text, const String& shortcutKeyText, const Drawable* drawableIcon, const Colour* textColourPtr) override
    {
        // Customize the appearance of each PopupMenu item here

        // For example, set the text color to red if the item is highlighted
        if (isHighlighted)
            g.setColour(Colours::red);

        // Call the base class implementation to actually draw the item
        LookAndFeel_V2::drawPopupMenuItem(g, area, isSeparator, isActive, isHighlighted, isTicked, hasSubMenu, text, shortcutKeyText, drawableIcon, textColourPtr);
    }

    // ... other methods ...
};
```

and then it says:
"This is just a basic example. Depending on what exact customization you want (changing highlight background color, text color, etc.), you might need to modify other methods.

Please integrate this into your existing LookAndFeelBase class and let me know if you encounter any issues or if there are more specific customizations you want to achieve."

so i dunno if that is helpful. i'm going to see if i can get it to address the specific issue now. ( i know, this is cheating, but .. )

**Reply by damiensellier** [reply]
in c++    &x is the pointer (the address)

with &g   go with  g.doSomething()  like g.fillAll() gsetColour()  g.setFont() etc depending on what's available for this component.

When doing LUA deep programming, I recommand to check the source code like we did to check the way the JUCE classes are implemented with LUA Bind. 

I 'll probably add it to Ctrlr if it doesn't crash. Sometimes in Ctrlr when you add a new feature it doesn't work and it takes hours to find the problem, because it's an old software that is still based on the older JUCE version and relies on fallback codes for legacy.
Good luck with your panel!

**Reply by subv0x** [reply]
i remember i used to look at both the JUCE and Ctrlr github - just getting a better idea of what was possible, at the time.

computer say:

"To modify the color of the PopupMenu background in the drawPopupMenuBackground method, you can use the g.fillAll() function to fill the entire area with a specific color. In the example I provided earlier, I used Colours::green to set the background color to green. You can replace it with any color you prefer.

Here's the relevant part of the code:"

```
void drawPopupMenuBackground(Graphics& g, int width, int height) override
{
    g.fillAll(Colours::green); // Set the background color of the PopupMenu to green
}
```
"If you want a different color, you can use other predefined colors from the Colours class or specify a color using RGB values. For example, to set the background to red, you can use:"
`
g.fillAll(Colour(0xFF, 0x00, 0x00)); // Red color`

**Reply by subv0x** [reply]
then asked it specifically for a method:

"If you're implementing this in Lua, you would be working with the LookAndFeel class through Lua bindings. Here's an example of how you might modify the drawPopupMenuBackground method in Lua:

lua

```
-- Assuming you have a custom LookAndFeel class
function CustomLookAndFeel:drawPopupMenuBackground(g, width, height)
    g:fillAll(Colours.green)
end
```

In this Lua script:
 CustomLookAndFeel is assumed to be your custom LookAndFeel class.
drawPopupMenuBackground is the method you're overriding.
g:fillAll(Colours.green) sets the background color to green.

Make sure that your Lua script is correctly integrated with your application and that it has access to the necessary JUCE classes.

Remember that this Lua script should be part of a larger setup where you set your custom LookAndFeel for your UI components. If you have a method to set the LookAndFeel in your application, you would set an instance of your CustomLookAndFeel there.

Also, in your LookAndFeel class, you might need to handle the Lua to C++ bindings. This might involve using a tool like "luabind" to expose C++ classes and methods to Lua. The exact process depends on your application framework and how it handles Lua scripting."

**Reply by subv0x** [reply]
so i've put this at the beginnng of my popup method, and it hasn't crashed. is there a check i can run to verify this 'custom class' ? i have no idea what i'm doing :P 

```
	-- Custom LookAndFeel for PopupMenu
	local CustomLookAndFeel = {}
	CustomLookAndFeel.new = function()
    	local self = {}
    	self.__index = self

    	setmetatable(self, {
        	__index = LookAndFeel
    	})

    	return self
	end
```

**Reply by damiensellier** [reply]
This chatgpt is incredible it's spot on!
it's just the drawPopupMenuItem is not bound to LUA (yet). I'll check that.
Once you have it bound, you can do lookandfeel() custom settings like you did with the popupBackground.

**Reply by subv0x** [reply]
uh didn't say i had it working yet. i'll have another go later.

**Reply by Ted-juh** [reply]
If you do a search at the Juce forums, you'll see more people having trouble with popupmenubackground. Therefore a newer popupMenuBackgroundWithOptions was introduced. But that one is hard to bind with Lua because it depends on a separate class.

I won't go in detail but the combobox, popupmenu and menubar are very close related which brought some confusion which kind of popupmenu you are altering. 

I tried it in a testpanel but I couldn't get it to work either. I don't like to brag but I'm quite adept with the whole l&f class. Lots of the l&f class works, some don't. But I got a nice purple background on the menubar now :)

**Reply by Ted-juh** [reply]
Sorry for the double post.

About chatgpt. Sometimes it's spot on, sometimes it's way off. Creating more trouble than helping you. Keep in mind that it's just copying bits of code from the web, which might or might not suit you. For the worse, be from someone who hadn't had a clue about Juce like we do from time to time.

**Reply by subv0x** [reply]
HOW do we change popup background, select highlight, and highlight text ?!
(was thinking a nice 50% alpha with luminous text ... )

**Reply by subv0x** [reply]
chatGPT, yes, can confuse and take you on detours. but it's instructive. and the level of interaction is, i would say, positive. you can chuck your brainjunk at it, ask it the same question repeatedly, ... it don't mind. although it will point out if you asked something before. it still needs a bit more tetchiness for that to be realistic, but it's a relief when you get a cheerful encouraging response.
chatGPT is good as a sounding board, and it's great how it will make suggestions, and even suggest sensible variable names. use it as a tool to chuck stuff at, and prompt yourself, as much as prompt it - ??? -

**Reply by subv0x** [reply]
however it might also make you inherently LAZY and less reliant on your best faculties.

**Reply by damiensellier** [reply]
@Ted-juh Are you refering to this thread? https://forum.juce.com/t/setting-comboboxs-popupmenu-colors/33057/7

**Reply by Ted-juh** [reply]
Been a while since I ran into the same problem subvox has.. but yes, looks like that is the thread I mentioned. And it states that the "PopupMenuBackgroundWithOptions" relies on the Options class, which isn't available through luabind yet if I remember correctly.

I could give it a try to add that class, since I (partially) succesfully added the NormalisableRange class. Then add the lookandfeel options for the PopupMenuBackGroundWithOptions def. Then test if it's possible to change the background color.

Don't expect it to be done by tomorrow.. I got a long list of things to do and I'm heavily invested into Cmake at the moment. But I'll give it a try.

**Reply by subv0x** [reply]
um, when you're both saying you'll 'add' a class, what are you referring to, working on a new version of Ctrlr? drafting your own Ctrlr? 

i do not know how to try the code chatGPT showed me. with the class creation thing, is that a separate method? does it have to be run by a panel Load method? what about the other bits? it looks like several (3) methods.

**Comment by subv0x** [comment]
ok- starting to begin to get my head around this ... or at least finding out where to look for conversations about it that can give some clues. 

thus, i glean: 
1/ i have to create the class with a method that gets called at startup. so lets call this 'initClass':

```
-- Initialize or set up the custom class
initClass = function()
    -- Custom LookAndFeel for PopupMenu
    CustomLookAndFeel = {}
    CustomLookAndFeel.new = function()
        local self = {}
        self.__index = self
        setmetatable(self, { __index = LookAndFeel })
        return self
    end
end
```

so now in theory i can have `CustomLookAndFeel.new()` - where ..? '.new' could be eg: 
`CustomLookAndFeel:drawPopupMenuItem(g, text, image, isSeparator, width, height, isHighlighted, isSelected)`

closest thing is:  `drawPopupMenuBackground()` from JUCE PopupMenu::LookAndFeel, for which parameters are `Graphics & , int  width,  int height ` - don't know if this gives access to anything other than width and height?
chatGPT suggested: 
```
	-- Create a function to draw the PopupMenu background with a custom color
	function drawCustomPopupMenuBackground(g, width, height, color)
    g:fillAll(Colour(color))
	end
```
don't know if that is valid,
 and still don't know how to build the function to be called from a button to test it - say a simple popup with 4 items, and try to make the background green FF00FF00 or something. 

( ... and now i'm seeing a JUCE discussion about `IsOpaque()` ... :-/ stumbled across a discussion about making transparent popup windows)

**Comment by subv0x** [comment]
been diggin' a bit over at JUCE forum, searching for topics with 'popupMenu' and 'background'. ('what i'm studying tonight')
came across several, and people seem to be getting their colours and transparency with LookAndFeel in JUCE. here's a first one, with some promising references:
https://forum.juce.com/t/bug-popup-menu-background-and-logic-silicon/43243/3
(example)
```
class TestLookAndFeel    : public LookAndFeel_V4
    {
    public:
        TestLookAndFeel()
        {
            const Colour newBgColour(53,53,53);
            const Colour newFgColour(220,220,220);
            setColour (PopupMenu::highlightedBackgroundColourId, newFgColour);
            setColour (PopupMenu::highlightedTextColourId, newBgColour.withAlpha(0.9f));
            setColour (PopupMenu::backgroundColourId, newBgColour.withAlpha(0.9f));
            setColour (PopupMenu::textColourId, newFgColour);
        }
        
        ~TestLookAndFeel() {}
        
        void drawPopupMenuBackground (Graphics& g, int width, int height) override
        {
            Rectangle<int> area(0, 0, width, height); //getLocalBounds().toFloat().reduced (0.5f);
            
            g.setColour(findColour (PopupMenu::backgroundColourId));
            //g.fillAll (findColour (PopupMenu::backgroundColourId));
            g.fillRoundedRectangle(area.toFloat().reduced(0.5f), 4.0f);
            //ignoreUnused (width, height);
        }
    };
```
which all looks quite familiar. or here, something else, about 'opaqueness', and something about a default white background default for 'opaque': (these are in no particular order)
https://forum.juce.com/t/menu-fillall-request/13231/2

another one on that:
https://forum.juce.com/t/white-background-of-popup-menu/9958

more on transparency:
https://forum.juce.com/t/popupmenu-transparent/48455/2

more chats on LookAndFeel:
https://forum.juce.com/t/several-popupmenu-look-and-feel/48316/6

oh look more, and now about Font:
https://forum.juce.com/t/popupmenu-fonts-using-lookandfeel/3400

this is what PopupMenu::Options looks like:
https://docs.juce.com/master/classPopupMenu_1_1Options.html

(nope, not there. wtf... why is this so hard)

lastly the original reference:
https://docs.juce.com/master/structPopupMenu_1_1LookAndFeelMethods.html#a65f8fd08fdf0a6c9d6fe7eafa1afb020

**Comment by Ted-juh** [comment]
Besides what you are trying to achieve with the Lua class, this is the way to do it in Ctrlr..

In the init file:

```
	lf = LookAndFeel()
	lf:setImplementation (pum1LF)
	panel:getComponent(nameOfTheComponent):setCustomLookAndFeel(lf)

```
Then add a new method that contains:
```
pum1LF = {
drawPopupMenuBackground = function (g, width, height) 
g.fillAll(Colour(0xFFFF0000))
end
}
```

But that doesn't work. And it's because of: panel:getComponent(nameOfTheComponent):setCustomLookAndFeel(lf). When you create the PopupMenu and name it "mainMenu", it doesn't work because it isn't a physical (?) component. Setting it to the button (PCM) won't matter because only the "button" l&F would apply.

One thing to try is to apply the lookAndFeel to the canvas because every component should take the L&F applied to it. But even then it's not applied to the PopupMenu.

Like I said last time.. some things are not bound/ linked right. And I can't tell if it's Juce or if it's Ctrlr (probably the latter). But when you add the scripts mentioned above and add a combo box... guess what? Yup, it has a red background in the drop-down menu. 

What I meant by "adding a class" is adding a class that binds the Juce "Options" class to Lua (Source/Lua/JuceClasses folder of the repository) in the luabind folder. It isn't there yet in Ctrlr, period. Writing your own Lua class/function won't help that much because you simply do not have access to the L&F class from Juce. But if the class is added to the JuceClasses, then it should be a little step to make use of "drawPopupMenuBackgroundWithOptions" which should offer a way to set the background of a popup menu. 

*Edit, lua class/function = the one you tried to write in posts mentioned before mine.

**Comment by subv0x** [comment]
@Ted-juh 
are you saying that what i want to do is not possible in Ctrlr? this is a whole new rabbit hole for me to go down (instead of finishing my panel...) and i'm getting my head around the general principles (vaguely, i have no proof of concept at my end). 

ChatGPT identified a lot of the issues i'm seeing mentioned. must've been hanging out at JUCE forums. 
i don't even really grasp how Lua and JUCE are related (didn't Lua come first, in the 60s?) (don't know where i saw that)
so i'm having a look at Damien's thread about L&F (heh getting the lingo) - it's ok: i get the idea that you can 'lua-bind' (although: why no upper case L there? ..) to JUCE in order to access parameters not directly available via Lua, and Ctrlr. i struggle to grasp how, if the guys developing JUCE decide to add something now, Ctrlr, finished in its current state several years ago, could access those changes in JUCE - unless Ctrlr is really built in JUCE? i'm just a brute force user, so show me the buttons, and what those knobs do, and i'll see what i can do. 

i would expect to be able to do this. i'm not the only person asking about it, it's on the JUCE forum. if i can conceive of a popup menu, i immediately think "oh what if that could be transparent with luminous text?' (like tomcruise in monetaryreport) - so we can only assume that there wasn't time for this. but i also see these being referred to as 'combos', which is the word we use for a specific modulator type. should i therefore forget about (very cool and easy) popup menu and use combo instead? no: you can't have submenus in Ctrlr combos. 
(sorry, waffling follow-up post). 

HOW? is there a 'how'? or forget about it?

**Reply by Ted-juh** [reply]
There might be a 'how' but not at this moment. So, better move on. Find another way. Get your thinking cap on. 

You can change the options provided in a combox. So if you have two buttons, let's name them pcm1 and pcm2, you can change the options of 1 combobox. Might that do the trick? 

For your info: Ctrlr IS made in Juce. You are basically building a panel inside a Juce project/ panel which you can export to different formats like standalone or vst. Quite genius if you ask me.

Lua is added to the mix as an easier programming language. But to get access to the Juce modules, there's luabind to do the translation to C++ in a way that Juce knows what class you need and which member of that class you need access to. But that's hard-coded into Ctrlr. I'm not entirely sure you can get access to those modules with a lua class, like you tried.

The reason those bindings still work is because Juce doesn't want to introduce changes that might break code. It happens, see Juce's change list at the repository. But Ctrlr 5.3.201 uses an older version of Juce and as long as it stays at that version, you won't get nasty bugs.

**Reply by damiensellier** [reply]
Popup full styling will be hard because it requires to setLookAndFeel() of the entire panel which we don't have access to.
You can change LookAndFeel at component level for a slider, button etc but a Popup is not really a component so the only way is to define its style from the ctrlrEditor.cpp. Maybe you could reach it with the callback on Canvas repaint "**Called to paint a custom panel Canvas**" in the globals. That would be where I'd put the popup lookandfeel definitions.

I'm working on the update of the entire Ctrlr software to get the V4 look and feel. I pulled my hair for hours about the popup on right clicking the canvas. Same for menuBar style. 
The way it will work on my next release is that you can select your "environnement" colourScheme (dark, midnight, light etc) and you should get the related popup colourScheme by default, same for menu, same for alertBox. But no real customization.

My advice about all that, is don't waste your time on this popup colour scheme for too long, it will never work entirely the way JUCE, Ctrlr and LUA are working all together. give a try to "Called to paint a custom panel Canvas" though.

**Reply by Ted-juh** [reply]
Nice work on the colour schemes and the l&f v4 implementation Damien. Been pulling my own hair over smaller things than you are trying to do.

Applying it to the canvas doesn't work either, by the way. As stated in an earlier post by me. I wonder what the difference is between when you apply a l&f to the canvas in a init file or the option you mentioned. Does one load earlier than the other so changes to the l&f do apply? 

Worth a shot.

**Reply by subv0x** [reply]
i've seen 'IsOpaque( )' in JUCE forum posts. same thing there?  if it could default to black (bodged transparent) instead of white ?
